core/src/main/java/io/questdb/cairo/sql/async/WorkStealingStrategyFactory.java

```diff
@@ -0,0 +1,41 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.sql.async;
+
+import io.questdb.cairo.CairoConfiguration;
+
+public class WorkStealingStrategyFactory {
+
+    private WorkStealingStrategyFactory() {
+    }
+
+    public static WorkStealingStrategy getInstance(CairoConfiguration configuration, int workerCount) {
+        final int noStealingThreshold = configuration.getSqlParallelWorkStealingThreshold();
+        if (workerCount >= 2 * noStealingThreshold) {
```

Comment:
> and the threshold be 1/2 of the worker count?

For large machines this will disable the feature. For instance, in case of 192 core box we have around 190 worker threads and 85 in-flight tasks is impossible to reach with the default queue sizes. So, the dependency is on both the number of worker threads and queue sizes and since there are multiple queues to consider, it'll make the default value too complex.

We could introduce an on/off switch in parallel with the current `cairo.sql.parallel.work.stealing.threshold` property and rely on it here. But its default logic will remain the same: `boolean enable = workerCount >= 2 * noStealingThreshold`. WDYT?

Reply:
I've changed the proportion to 4x instead of 2x since this feature is meant only for beefy machines: 6b7b466c00564b0acc86f6e9a7cd93f1caf8718c